### PR TITLE
Allow cash ticking sound to overlap

### PR DESCRIPTION
**Pull request by PunkPun**
Supersedes #20296

Implemented as suggested in https://github.com/OpenRA/OpenRA/pull/20296#issuecomment-1248222525

Suggestions for better naming are welcome

### CODE REVIEW

File: OpenRA.Game/GameRules/SoundInfo.cs [hunk 1]
```diff
@@ -48,10 +48,10 @@ public SoundInfo(MiniYaml y)
 				if (volumeModifierNode != null)
 					volumeModifier = FieldLoader.GetValue<float>(volumeModifierNode.Key, volumeModifierNode.Value.Value);
 
-				var allowInterrupt = false;
-				var allowInterruptNode = t.Value.Nodes.FirstOrDefault(x => x.Key == "AllowInterrupt");
+				var allowInterrupt = SoundPool.DefaultInterruptType;
+				var allowInterruptNode = t.Value.Nodes.FirstOrDefault(x => x.Key == "InteruptType");
```
Comment (abcdefg30) [inline OpenRA.Game/GameRules/SoundInfo.cs]:
```suggestion
				var allowInterruptNode = t.Value.Nodes.FirstOrDefault(x => x.Key == "InterruptType");
```
Can we use `nameof(SoundPool.InterruptType)` here?

File: OpenRA.Game/GameRules/SoundInfo.cs [hunk 2]
```diff
@@ -64,15 +64,17 @@ public SoundInfo(MiniYaml y)
 
 	public class SoundPool
 	{
+		public enum InterruptType { NeverInterrupt, AllowInterrupt, AllowOverlap }
+		public const InterruptType DefaultInterruptType = InterruptType.NeverInterrupt;
 		public readonly float VolumeModifier;
-		public readonly bool AllowInterrupt;
+		public readonly InterruptType InteruptType;
```
Comment (abcdefg30) [inline OpenRA.Game/GameRules/SoundInfo.cs]:
```suggestion
		public readonly InterruptType InterruptType;
```

File: OpenRA.Game/Sound/Sound.cs [hunk 3]
```diff
@@ -401,36 +401,42 @@ public float VideoVolume
...
-				if (sound == null)
-					return false;
 
-				if (voicedActor != null)
-					currentSounds[actorId] = sound;
-				else
-					currentNotifications[name] = sound;
+					var sound = PlaySound();
+					if (sound != null)
+						currentSounds[actorId] = sound;
```
Comment (abcdefg30) [inline OpenRA.Game/Sound/Sound.cs:445]: This is no longer returning false now if the sound is null.

File: OpenRA.Game/Sound/Sound.cs [hunk 4]
```diff
@@ -401,36 +401,42 @@ public float VideoVolume
...
-				if (currentNotifications.TryGetValue(name, out var currentNotification))
+				ISound PlaySound()
 				{
-					if (!currentNotification.Complete)
+					var volume = InternalSoundVolume * volumeModifier * pool.VolumeModifier;
+					return soundEngine.Play2D(sounds[name], false, relative, pos, volume, attenuateVolume);
+				}
+
+				if (pool.InteruptType == SoundPool.InterruptType.AllowOverlap)
+					PlaySound();
```
Comment (abcdefg30) [inline OpenRA.Game/Sound/Sound.cs]: Same here.

**Review by abcdefg30 — COMMENTED** [review]
This also needs an update rule.

File: OpenRA.Game/GameRules/SoundInfo.cs (hunk 1 above)
Comment (PunkPun) [reply OpenRA.Game/GameRules/SoundInfo.cs]: sure

File: OpenRA.Game/GameRules/SoundInfo.cs (hunk 2 above)
Comment (PunkPun) [reply OpenRA.Game/GameRules/SoundInfo.cs]: fixed it to `Type`

**Comment by PunkPun** [comment]
> This also needs an update rule.

We don't have updates rules for Notifications

**Comment by abcdefg30** [comment]
Ah right, never mind then.

**Comment by anvilvapre** [comment]
Regarding terminology, `InterruptType`, `OverlapType`, `MaxParallelVoices` assuming the interrupting only relates starting to play again the same sound.

But the first seems fine, max sense.

**Comment by PunkPun** [comment]
This relates to what we do with our new sound if a sound is already playing. As in do we not play it, do we interrupt the old sound and then play ours, and lastly do we not care and just let them overlap

**Review by abcdefg30 — APPROVED** [review]

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9dd28f68d661f98b7dde458dad48700ae63304a7...5a2b3f5d0eb2a7f12a8f5d1a9affa245d6e5ea78)

**Pull request merged**